Feature: Endpoint, Models, Routes, Controller Development

gaal1le0 opened the pull request:
This PR contains all data staff needed to perform server request to the API

Commit:
Added more fields and refactor ones

Commit:
Refactor some code

Commit:
Created database

Commit:
Added direction field

Commit:
Include direction object

Commit:
Included direction in the action to transform objects

Commit:
Format code

Commit:
Create validator of jwt

Commit:
Verified token

Commit:
Create middelware to validate JWT Token

Commit:
Added documentation

Commit:
Documented and added Sign Token middelware

Commit:
Added JWT Token

Commit:
Created token payload and implement actor to decode data

Commit:
Format documment

Commit:
Compute general exp time

Commit:
Implemented debugger modules and rename files

Commit:
Create debugger env and types

Commit:
Debugger standard methods

Commit:
Implemented debugger protocol

Commit:
Finished complete debugger to acomplish different environments

Commit:
Manage security of API

Commit:
Fixed bugs with the names

Commit:
Minor improvements

Commit:
Create relationship between accounts and dependent

Commit:
Implemented distribution in files

Commit:
Renamed files

Commit:
Renamed files again

Commit:
Reorganize project

Commit:
Implemented login workflow in the API and refactor some interfaces

Commit:
Created endpoint to delete account by ID

Commit:
Designed endpoint to update account status

Commit:
Implemented access control and granted access in some endpoitns

Commit:
Securized all the endpoints

Commit:
Fixed problems with Request. Setted a global extension type

Commit:
Renamed all files

Commit:
Changed indotem SQL files

Commit:
Reorganazie SQL Folder

Commit:
Restructure file names

Commit:
Installed types of some libraries

Commit:
Alter some references

Commit:
Included more documentation

Commit:
Fixed some errors

Commit:
Setting routes

Commit:
Created the endpoint to update account data

Commit:
Setted routes

Commit:
Created endpoint to update password

Commit:
Remove unnecesary imports

Commit:
Setted global environment debugger

Commit:
Expose debugger via Singleton

Commit:
Started validator

DanielBanobreDopico committed:
Verify permissions for foreign accounts

DanBaDo commented on src/routes/routes.accounts.ts:
```diff
@@ -0,0 +1,59 @@
+   * @param acccountID {Number}
+   * @protected Admin or User with accountID
+   */
+  app.get(`/:accountID`, AccountsController._getById);
```
We can simplify access control by dividing this endpoint in two:

- /me/
   - Without parameters
   - Provides only data related to the authentication token account.
-  /accounts/:accountID/
   - Granted only for Admin
   - Provides data for parameter accout.
- /dependents/
   - Provides only dependents related to the authentication token account.
- /dependents/:accountID/
   - Granted only for Admin
   - Provides dependents related to the parameter account.

... and so.

gaal1le0 replied:
Why? 
What do you think about it? Do you want to make a call to discuss it?

gaal1le0 replied:
Change approved :)

Commit:
Dependent model interface

Commit:
Get account dependents controller

Commit:
Get account dependants route

Commit:
Fix constant name typo

Commit:
Account - dependent relationship table name constant

Commit:
Find all account dependets action

Commit:
Fix table name

Commit:
Add documentation

DanBaDo commented on src/http/common/Debug/Debugger.ts:
```diff
@@ -0,0 +1,61 @@
+   */
+  _manageLogType = (type: DebuggerTypes, log: string) => {
+    switch (type) {
+      case DebuggerTypes.alert:
```
It looks like a typo: alert/warning

gaal1le0 replied:
Yes but in the future we will wide this types.

gaal1le0 commented:
@DanBaDo I have just approved your changes. It's okey.
Only we should discuss about the endpoints you said me in a comment before. Contact me to schedule a call today ;)

gaal1le0 commented on src/routes/routes.accounts.ts:
```diff
@@ -0,0 +1,65 @@
+   * @protected Admin or User with Account ID
+   * @param accountID
+   */
+  app.put('/:accountID/dependents', AccountsController._getDependents);
```
This is not a `PUT` method, it's only a GET

gaal1le0 commented:
@DanBaDo @Alcazar87 you must separate all data staff before creating new endpoints. You have mixed all the staff related with dependents inside de `controller.accounts.ts` file, you must create another file for routing, actions, controller for the `depentents`. If you put all the staff code inside the accounts files, it's coupling.

gaal1le0 commented:
@DanBaDo @Alcazar87 the endpoints that are needed to be created in `routes.dependents.ts` are:
- Get all dependents asscociated with an account ID
- Alter get all accounts endpoint to extract all related dependents and construct a DTO object (ask me about)
- Get dependent asscociated with an accountID and dependentID
- Save new dependent associated to an accountID (one or many accountID)
- Alter dependent status (PATCH)
- Alter all dependent data (PUT) sql query should be created dinamically
- Delete dependent associated with dependent ID and accountID (given by the token) -> user access
- Delete dependent by ID
- Delete all dependent associated to an accountID -> user access

gaal1le0 commented:
> @DanBaDo @Alcazar87 you must separate all data staff before creating new endpoints. You have mixed all the staff related with dependents inside de `controller.accounts.ts` file, you must create another file for routing, actions, controller for the `depentents`. If you put all the staff code inside the accounts files, it's coupling.

The main idea for this change is focusing on create UNIT test dedicated to each data staff following TDD principles. If not the test will be overcomplicated and we will violated all the SOLID principales, mainly the: "Unique responsability"

Commit:
Reorganize files

Commit:
FEATURE: Access granted permission validator (Middelware) 

* Setted basic structure
* Create initial structure of the middelware
* Create function to run validations
* Refactor function
* Documented function
* TS Format applied
* Reorganize files

Commit:
FEATURE: Query Constructor Data Staff 

* Move files structure

* Create needed files

* Create basic classes

* Creaet more structure

* Create data staff structure

* Added more based structure

* Create query runner and adapted query builder

* Documented query runner

* Remove unncesary imports

* Implemented all methods in query builder

* Bugs fixing and implemented internal methods

* Create basic table structure

* Creted new types and refactor code

* Set getter method and init data staff

* Created helper and protocols

* Resolve some problems

* Abstracted helper

* Renamed object and refactor code

* Refactor code and implement save method

* Refactor object to enums

* Added documentation to code constructor

* Created update constructor

* Refactor some code

* Created delete constructor

* Created query builder runner

* Change access control

Commit:
Fix route method

Commit:
Split account and dependents code

gaal1le0 reviewed (commented):
Perfect @DanBaDo. All the changes has been approved.

DanBaDo commented:
> > @DanBaDo @Alcazar87 you must separate all data staff before creating new endpoints. You have mixed all the staff related with dependents inside de `controller.accounts.ts` file, you must create another file for routing, actions, controller for the `depentents`. If you put all the staff code inside the accounts files, it's coupling.
> 
> The main idea for this change is focusing on create UNIT test dedicated to each data staff following TDD principles. If not the test will be overcomplicated and we will violated all the SOLID principales, mainly the: "Unique responsability"



> @DanBaDo @Alcazar87 you must separate all data staff before creating new endpoints. You have mixed all the staff related with dependents inside de `controller.accounts.ts` file, you must create another file for routing, actions, controller for the `depentents`. If you put all the staff code inside the accounts files, it's coupling.

Fixed!

Commit:
Add dependents resources and permissions

Commit:
Chage status from 500 to 401

Commit:
Use AccessControlMiddleware

gaal1le0 reviewed (commented):
Ok @DanBaDo

DanielBanobreDopico committed:
Imports cleanup

DanBaDo commented on src/http/controllers/controller.dependents.ts:
```diff
@@ -15,42 +11,27 @@ export module DependentsController {
-            if (!accountID) {
+        const { accountID } =
+            req.user.role == PermissionRoles.USER ? req.user : req.params;
+        if (!accountID) {
```
AccessControlMiddleware must grants runs this callback  only to authenticated accounts . We don't need verify if we get a accountID becouse we alwais will get it.

gaal1le0 replied:
Yes it's right only you need to use `req.params`. But are you sure dont you need to check if it's null?

gaal1le0 replied:
This is the problem I mentioned you that express didn't recognize the differente between `dependents` & `depdents/{id}`. If u want we can make a call to take more into detail about it.

DanBaDo replied:
I splited it in tow controllers.
Non ADMIN queries from myDependents/:dependentID/ uses _getMyDependentById in line 43. This controller returns one dependent  if a valid dependentID is provided or all account dependents if no dependentID is provided.
ADMIN quieries from  /dependentsFor/:accountID uses _getDependents in line 13.

gaal1le0 replied:
Please remove all this validations, it's not needed more due to our new implementation.

DanielBanobreDopico committed:
New route myDependents/{dependentID}/

DanielBanobreDopico committed:
New controller _getMyDependentById

DanielBanobreDopico committed:
Add documentation to controller

DanielBanobreDopico committed:
Fix argument lack

DanielBanobreDopico committed:
New action findOneAccountOwnedDependantById

Commit:
Add action for no dependentID

Commit:
Rename action

Commit:
New route POST /myDependents/

Commit:
New controller _newOwnedDependent

Commit:
New action insertNewDependent

Commit:
Fix row name in query string

Commit:
Fix dependand->dependent

Commit:
Add map to findDependentByOwnerAndID

Alcazar87 committed:
Replace account for customer

gaal1le0 commented:
@DanBaDo tell me when all the changes are made.

Alcazar87 committed:
Fix some bugs

Fixed some names  with uppercase letter

Alcazar87 committed:
Created controller.auth

Created controller.auth and reorganize _login and _updatePassword

Alcazar87 committed:
add export AuthController

Alcazar87 committed:
create new route: auth

create routes.auth and reorganize /login and /password

Alcazar87 committed:
create actions.auth

Create actions.auth and reorganize login and updatePassword from actions.customers

Alcazar87 committed:
Replace CustomerActions for AuthActions in controllers

gaal1le0 reviewed (commented):
@Alcazar87 all your changes has been approved

DanielBanobreDopico committed:
Fix insertNewDependent exit and cleanup

DanielBanobreDopico committed:
Remove paths camelCase

DanielBanobreDopico committed:
Fix route-controller calls

Commit:
Include runner example

gaal1le0 commented:
@DanBaDo your lasts changes has been approved

Commit:
Customer DB runner work in progress

DanBaDo commented on src/http/actions/actions.customers.ts:
```diff
@@ -44,21 +44,22 @@ export module CustomerActions {
-    db.query(queryString, (err, result) => {
-      if (err) {
-        callback(err);
+    _runner.run(queryString, (res: SQLQueryResponse<Array<TinyCustomer>>)=>{
```
I found that I need to define a type for responses in actions if I whant to use res.map() for parse directions JSON, but when I do it can't parse JSON becouse item.direction is a Direction not a string. How must I solve it?.

DanBaDo replied:
I'v tried to use ```SQLQueryResponse<Array<Object>>``` instead, but then can't use item properties in map becouse "Object type has not bla,bla property"

gaal1le0 replied:
Let's going to make a call

Commit:
Add customer DTO interface

Commit:
Apply SQLrunner to findAll action

gaal1le0 commented on src/models/types/model.customer.ts:
```diff
@@ -19,4 +19,16 @@ export interface UpdateCustomerModel {
+export interface CustomerDTO {
+  id?: Number;
+  name?: string;
+  username: string;
+  email: string;
+  phone: string;
+  direction: string;
+  password: string;
+  status: Number;
+  date: number;
```
All these fields must be optional, because when you perform a server request you will get different values in each request. Make all this staff optional, if not your code will not work when we make filter server request. 
If you are not sure what does it mean please ask me.

gaal1le0 replied:
@DanBaDo @Alcazar87 This is **REALLY** important. Please both of you check this staff.

Alcazar87 committed:
Change fields to optional

DanBaDo commented:
Ok

DanielBanobreDopico committed:
Comments cleanup

DanielBanobreDopico committed:
Applying SQLrunner to customers actions

DanBaDo commented on src/http/actions/actions.customers.ts:
```diff
@@ -73,10 +73,10 @@ export module CustomerActions {
-        callback(err);
+    const queryString = `DELETE FROM ${_tableName} WHERE id=${customerID}`;
+    _runner.run(queryString, (res)=>{
+      if (res.err) {
```
We are using ```if``` structure calling ```callback```, and calling ```callback``` again without ```else```. I'm afraid it can produce problems in the future if ```callback``` returns control to the current code and is called twice.

gaal1le0 replied:
Callback is a function that returns `Void` then the code that if outside of the if will never execute if it enter inside the `if` block. Don't worry about that, it's not a problem. For now it's the best approach to our solution, if in the future we need to return a no void function, we will change it later, but actually the runner has no meaning to have an else clause inside. If you are not sure to understand what does it mean, please contact directly with me.

DanielBanobreDopico committed:
Minor code visual improvement

DanielBanobreDopico committed:
Using SQLRunner table replacement

Commit:
SQL responses types

Commit:
New dependents DTOs

Commit:
Add newDependent and linkDependentToCustomer actions

Commit:
Add _newDependentForCustomer controller

Commit:
Add new POST /dependents/ endpoint

Commit:
Documentation

Alcazar87 committed:
Add DependentDTO

Alcazar87 committed:
Applying SQLrunner to Auth actions

Alcazar87 committed:
Applying SQLrunner to dependents actions

Alcazar87 commented on src/http/actions/actions.dependents.ts:
```diff
@@ -113,31 +116,32 @@ export module DependentsActions {
       } else {
         // insert relation
-        const newDependentId = result.insertId;
+        const row = res.data[0];
```
I´m not 100% sure that we receive the correct information in this response. Please check it

DanBaDo commented:
All right

Commit:
Merge

Commit:
New POST /dependents/ endpoint

Commit:
New action getDependentById

Commit:
New controller _getDependentById

Commit:
New endpoint GET /dependent/:dependentID/

Commit:
Add dependents permisions to owners

Commit:
Apply AccessControlMiddleware to auth routes

Commit:
Remove auth _updatePassword controller deprecated access control

DanBaDo commented on src/http/actions/actions.dependents.ts:
```diff
@@ -11,10 +11,12 @@ import {
-const _runner = new SQLRunner(db, _tableName);
+// * SQL Runners to perform MYSQL Requests
+const _dependentsRunner = new SQLRunner(db, _dependentTableName);
+const _relationshipRunner = new SQLRunner(db, _customerDependentRealtionshipTableName);
```
We work mailly with dependents table here and with relationship table too.
Using SQLRunners for both tables reduces the SQL query chains length.

DanBaDo replied:
@Alcazar87 please check if it's ok for you.

DanBaDo commented on src/http/actions/actions.dependents.ts:
```diff
@@ -159,4 +161,67 @@ export module DependentsActions {
+   * @param direction {string}
+   * @param callback {Function}
+   */
+  export const newDependent = (
+    name: string,
+    phone: string,
+    direction: string,
+    callback: Function
+  ) => {
```
This action and the next one can replace the previous ```insertNewDependent``` action in a simplest and reusable way.

DanBaDo commented on src/http/controllers/controller.dependents.ts:
```diff
@@ -91,4 +91,70 @@ export module DependentsController {
+                            (err)=>{
+                                if (err) {
+                                    _handleResponse(
+                                        {
```
I would like to avoid nesting ```.linkDependentToCustomer``` in ```.newDependent``` if it's possible.

gaal1le0 replied:
It's for me or for @Alcazar87 ?

Alcazar87 commented:
Sounds good to me. I understand so much better how runner works and how to build it after your changes

Commit:
Cleanup

gaal1le0 reviewed (commented):
Okey, approved all previous changes

Commit:
Applying AccessControlMiddleware to some endpoints

Alcazar87 committed:
New endpoint deleteDependent

Alcazar87 committed:
Add deleteDependent to controller.dependents

Alcazar87 committed:
new endpoint DELETE /dependents/:dependentID/

Alcazar87 committed:
new endpoint DELETE /my-dependents/:dependentID/

Alcazar87 committed:
Fix some bugs

Alcazar87 commented on src/routes/routes.dependents.ts:
```diff
@@ -64,7 +64,7 @@ export const DependentsRoute = (app: express.Application) => {
    */
      app.delete('/dependents/:dependentID', AccessControlMiddelware._grantAccess((query) => {
-      return query.readAny(_resource)
+      return query.deleteAny(_resource)
```
I´ve replaced readAny for deleteAny because it makes more sense to me, but i don´t know if it is right, because in customers @whitehatdevv used readAny in the endpoint "DELETE('/:customerID". Please check it.

Alcazar87 replied:
@whitehatdevv  this comment

gaal1le0 reviewed (commented):
All this changes are approved.

Alcazar87 committed:
Fix a bug

gaal1le0 reviewed (commented):
This changed was approved. Really just good work 🤩

Commit:
Some fixes

Commit:
Apply AccessControlMiddelware to customers endpoints

DanBaDo commented on src/http/actions/actions.dependents.ts:
```diff
@@ -250,4 +250,24 @@ export module DependentsActions {
+   */
+    export const deleteDependent = (dependentID: string, callback: Function) => {
+      const queryString = `
+      DELETE 
+      FROM ${_dependentTableName}
+      INNER JOIN ${_customerDependentRealtionshipTableName} 
+      ON ${_dependentTableName}.id = ${_customerDependentRealtionshipTableName}.dependentID 
+      WHERE id=${dependentID}`;
```
@Alcazar87 This action is used by '/my-dependents/:dependentID' endpoint who must grant only to delete owned dependents for loged in customers.
No controler or actions verifies that dependent is owned by cuntomer calling endpoint, so any logged customer can delete any dependent.
I think we must write two separated "chains" endpoint -> controller -> action for admins and customers. The exisitng action is good for admins with a ".deleteAny" permission, but we need a new one for customers with ".deleteOwn" permissions.

Alcazar87 committed:
New action deleteRelationshipUponDependet

Alcazar87 committed:
Add deleteRelationshipUponDependent to deleteDependent controller

DanielBanobreDopico committed:
Some fixes for dependent deletion

Commit:
New route POST /account/

Commit:
New controller _newCustomer

Commit:
New action insertNewCustomer

Commit:
Merge branch 'feature/endpoint-development' of github.com:Indotem-Salddud/account-managenment-api into feature/endpoint-development

gaal1le0 commented on src/routes/routes.customers.ts:
```diff
@@ -54,4 +54,8 @@ export const CustomersRoute = (app: express.Application) => {
     return query.updateOwn(_resource)
   }), CustomersController._updateById);
+
+  app.post('/account/', AccessControlMiddelware._grantAccess((query) => {
```
What is that?

gaal1le0 replied:
This endpoint should be called `onboarding/`. 
Please remove the controller right now. The process is really overcomplicated, you haven't take care into the most important detail

DanBaDo replied:
It's the administrative new customer account creation.

DanBaDo replied:
Ok. I've readed next comments. I remove it too.

gaal1le0 commented on src/http/controllers/controller.customers.ts:
```diff
@@ -183,4 +184,38 @@ export module CustomersController {
+   * @param req {Request}
+   * @param res {Response}
+   */
+  export const _newCustomer = async (
+      req: express.Request,
+      res: express.Response
+    ) => {
+    // get customer data
+    const { name, username, email, phone, direction, password } = req.body;
+    // call to action
+    CustomerActions.insertNewCustomer(
+      name,
+      username,
+      email,
+      phone,
+      direction,
+      password,
+      (err?: string, data?: TinyCustomer) => {
+        if (err) {
+          _handleResponse(
+            { statusCode: 500, message: 'Server response cannot be processed' },
+            res
+          );
```
All this function from `controllers` must be removed. This will be the onboarding and it's the most complicated one, we will perform this stuff together.

DanBaDo replied:
Ok.

gaal1le0 commented on src/http/actions/actions.customers.ts:
```diff
@@ -131,4 +133,49 @@ export module CustomerActions {
+   * @param password {string}
+   * @param callback {Function}
+   */
+  export const insertNewCustomer = (
```
The same as before, remove this action

gaal1le0 commented:
@DanBaDo check the latest comments, if you have any question ask me directly and we can meet again. 
@Alcazar87 you should check this stuff too, it's really important. 
Onboarding is the most complicated process in `account-managenemnt` API

Commit:
Remove new customer insertion

gaal1le0 commented:
@DanBaDo have you just understand our changes?

Alcazar87 committed:
DeleteDependent Controller reorganize

Alcazar87 committed:
Some changes in deleteDependent

gaal1le0 reviewed (commented):
All changes approved

Alcazar87 committed:
new UpdateDependentModel interface

Alcazar87 committed:
new updateDependentData action

Alcazar87 committed:
new _updateDependendentById controller

Alcazar87 committed:
new 'PUT dependents/:dependentID' and 'PUT /my-dependents/:dependentID' routes

gaal1le0 commented on src/models/types/model.dependent.ts:
```diff
@@ -20,6 +20,17 @@ export interface newDependentForCustomerDTO {
+ * * Alcazar87 - 2022/01/03
+ */
+
+export interface UpdateDependentModel {
+  name?: string;
+  phone?: string;
+  direction?: Direction;
+}
```
Why have you used this model to update a dependent?

DanBaDo commented on src/http/actions/actions.dependents.ts:
```diff
@@ -0,0 +1,340 @@
+    const queryString = `
+      UPDATE ${_dependentTableName}
+      SET
+        name=IsNULL(@name, ${updatedData.name}),
+        phone=IsNULL(@phone, ${updatedData.phone}),
+        direction=IsNULL(@direction, ${updatedData.direction})
```
I think we can't use @ SQL substitution with SQLRunner. We have the JavaScript ${} substitution instead.

gaal1le0 replied:
Of course @DanBaDo this change is approved, please proceed to change all the staff to use these SQL substitutions.

Alcazar87 committed:
new updateDependentStatus action

Alcazar87 committed:
new updateDependentStatus controller

Alcazar87 committed:
new  'PATCH /status/:dependentID' route

gaal1le0 reviewed (commented):
Changes has been approved

Alcazar87 committed:
Divide status/:customerID  user and admin routes

Renaming routes is still pending

Alcazar87 committed:
new updateOwnDependentStatus action

Alcazar87 committed:
new UpdateOwnDependentStatus controller

Alcazar87 committed:
new /myStatus/:dependentID route

Renaming routes is still pending

DanBaDo commented:
It looks all ok for me

Commit:
New customer endpoint GET /myAccount/

Commit:
New customer action _getOwnCustomerAccount

Commit:
Some path updates and a permission fix

Commit:
New customer endpoint PUT /customers/:customerID

Commit:
New customer controller _updateOwnCustomerAccount

Commit:
Fix identifications problems and split one endpoint

Commit:
New endpoint GET /dependentOwners/:depedentID/

Commit:
New dependents controller _getOwnersByDependentId

Commit:
New dependents action getOwnersByDependentId

Commit:
Trivial fix

Commit:
Fix access control resources

Commit:
Actions headers and DB models for refres JWT

Commit:
Fix functions names

gaal1le0 commented on src/models/sql/sql.jwt.sql:
```diff
@@ -0,0 +1,9 @@
+-- Contains JWT refresh tokens
+CREATE TABLE `JWTRefreshTokens` IF NOT EXISTS (
+    customerID INT PRIMARY KEY NOT NULL,
+    refreshToken BINARY(32), /*TODO: is 256 bytes ok?*/
+    expiration TIMESTAMP,
+    FOREIGN KEY (customerID) REFERENCE customers(id),
+) ENGINE = InnoDB,
+CHARSET = utf8,
+COLLATE = utf8_spanish2_ci;
```
What is that?
Why have you created a table to store all the refresh tokens?

DanBaDo replied:
We need a backend shared storage for refresh-tokens for granting all nodes access them.
Alternatively, we can use self-contained non-stored signed tokens, like authentication tokens are, but we will lost control over they if we need to revocate some of them.
I suppose it's possible to use on-memory tables or something like Redis for let all service nodes know about refresh-tokens state, but refresh tokens must not have a intensive use and DB access must not suppose a big work load accessing DB.

gaal1le0 replied:
No, it's not our approach, we don't need to storage any refresh token, that's the key to success. Here we are focusing only on disable accounts which don't need to sign in again, therefore (thanks to JWT middleware second step which checks if accounts is enabled) allow us to disabled all refresh tokens without having to storage it and without using any third part library. 

If you are not sure to understand what I'm proposing please ask me directly.

Commit:
Setted difference between app and server

Commit:
Remove refresh token db model and actions

Commit:
Merge branch 'feature/endpoint-development' of github.com:Indotem-Salddud/account-managenment-api into feature/endpoint-development

Commit:
Rename some routes

Commit:
new profile action

isOnBoarding is still pending

Commit:
new _profile controller

Commit:
new profile route and rename some routes(minor changes)

Commit:
Merge branch 'feature/endpoint-development' of https://github.com/Indotem-Salddud/account-managenment-api into feature/endpoint-development

Alcazar87 committed:
Rename some routes

gaal1le0 commented:
@Alcazar87 @DanBaDo All changes has been approved

Commit:
Fixed missing export and imports

Commit:
Adds cascade delete to customers deletion

Commit:
Adds customer cascade deletion

Commit:
Create response handler and edit send method

Commit:
Fixed middelware responses

Commit:
Fixed customer controller

Commit:
Fixed dependents response

Commit:
Remove unnused imports

DanielBanobreDopico committed:
Fixes indentation

DanBaDo reviewed (commented):
ok

Commit:
Global properties and error handler generator

DanielBanobreDopico committed:
Change error type to Number

DanielBanobreDopico committed:
Merge branch 'feature/endpoint-development' of github.com:Indotem-Salddud/account-managenment-api into feature/endpoint-development

gaal1le0 commented:
@DanBaDo your changes have been approved.

Commit:
Added more documentation to the error handler

Commit:
Setted the global date

Commit:
Remove hard string on endpoints and abstracts all available ones

Commit:
Create translator key interface

Commit:
Created basic translator keys

gaal1le0 commented:
> #### ⚠️ GitGuardian has uncovered 1 secret following the scan of your pull request.
> Please consider investigating the findings and remediating the incidents. Failure to do so may lead to compromising the associated services or software components.
> 
> 🔎 Details of the secret
> 🛠 Guidelines to remediate hardcoded secrets
> ###### [GitGuardian](https://dashboard.gitguardian.com/auth/login/?utm_medium=checkruns&utm_source=github&utm_campaign=cr1) is an automated secrets detection service.We help developers and security teams secure the modern software development process.
>  
> 
> ###### Our GitHub checks need improvements? [Share your feedbacks](https://qxhaizuqzpk.typeform.com/to/KmeAPTMk)

@DanBaDo @Alcazar87 please research more about this error ASAP and give me an update about it. Give me the update replying to this comment.

Commit:
Fix SQL query

Commit:
Adds action to remove dependents that haven't customers

Commit:
Merge commits fixing SQL query

DanBaDo commented:
I undestand it's not a security fault. It looks like GitGuardian found so many security related words in near value asignment (key, login, auth) and undestand the UUID is a secret key, when it's not.

gaal1le0 commented:
> I undestand it's not a security fault. It looks like GitGuardian found so many security related words in near value asignment (key, login, auth) and undestand the UUID is a secret key, when it's not.

@DanBaDo Then we should refactor the names to avoid this error because all the check must be successful, please refactor all the names and tell me know if you have any question.

DanielBanobreDopico committed:
Rename Tanslator keys

DanielBanobreDopico committed:
Applies ggignore

DanielBanobreDopico committed:
Adds new ggignore

DanBaDo committed:
ADD: JWT refresh architecture token to login

Co-authored-by: Daniel Banore Dopico <daniel.banobre.dopico@pm.me>

DanielBanobreDopico committed:
Add cascade deletion to customer deletion.

gaal1le0 reviewed (commented):
Changes approved

DanBaDo committed:
BUG: Add ignored paths to GitGuardian 

Co-authored-by: Daniel Banore Dopico <daniel.banobre.dopico@pm.me>

DanielBanobreDopico committed:
Add new keys and UUIDs

DanielBanobreDopico committed:
Fix label

DanielBanobreDopico committed:
Merge imports

Commit:
Added verification middelware to all needed endpoints

DanielBanobreDopico committed:
Merge branch 'feature/endpoint-development' of github.com:Indotem-Salddud/account-managenment-api into feature/endpoint-development

DanielBanobreDopico committed:
Add error handler and translatios to auth controller

DanielBanobreDopico committed:
Remove invalid translation key

Commit:
New action newCustomer

Commit:
Revert "New action newCustomer"

This reverts commit b0f9f6d9fb4a25105eb054871dc1d1745903075d.

Commit:
TranslatorKeys and TranslatorKeyUUIDs added

gaal1le0 reviewed (commented):
Changes approved

Commit:
Login controller updated

Commit:
Create SuccessHandler

Commit:
Fix login controller

gaal1le0 reviewed (commented):
Changes approved.

Commit:
Update updatePassword controller

Commit:
Fixes export name

Commit:
Adds lost routes

Commit:
Merge branch 'feature/endpoint-development' of github.com:Indotem-Salddud/account-managenment-api into feature/endpoint-development

Commit:
Update findByID action

Commit:
Update getOwnCustomerAccount y getAll controllers(customers)

Commit:
Fix some errors

Commit:
Update deleteCustomer controller

Commit:
Minor Change

Commit:
Update updateStatusById controller

Commit:
Update updateOwnStatus controller

Commit:
Update updateById controller

Commit:
Update updateOwnCustomerAccount controller

Commit:
Update profile controller

Commit:
Minos Change

Commit:
Update getMyDependents controller

Commit:
Update getAllCustomerDependents controller

Commit:
Update getMyDependentById controller

Commit:
Update newOwnedDependent controller

DanBaDo commented:
It looks ok for me.

Commit:
Update  getDependentById and deleteDependent controllers

Commit:
Update updateDependentById controller

Commit:
Update updateDependentStatus controller

Commit:
Update updateOwnDependentStatus controller

Commit:
Update getOwnersByDependentId controller

Commit:
Method removed from keys and some new success keys added

Commit:
Some errors fixed

Commit:
Fix an error

Commit:
Simplifies dependents insertion

Commit:
Merge branch 'feature/endpoint-development' of github.com:Indotem-Salddud/account-managenment-api into feature/endpoint-development

Commit:
Adds 203 message for incomplete dependent creation

Commit:
Update messages in _newDependentForCustomer controller

Commit:
Fix error code

Commit:
New dependentsSoleForCustomerID action

Commit:
Add dependentsSoleForCustomerID endpoint

Commit:
Add dependentsSoleForCustomerID keys and uuids

Commit:
Add dependentsSoleForCustomerID controller

Commit:
New GetDependentsSoleForCustomerID route

Commit:
Fix ambiguous endpoint string in error messages

DanBaDo commented:
@whitehatdevv @Alcazar87 @StefAranzazu 
Commit 7ff7a7b proposes a solution for ambiguous endpoint strings in errors for controllers called from multiple routes. Please, review.

Commit:
Removes unused import

Commit:
Replace req.baseUrl by req.route.path
